### PR TITLE
Make more code conform to strict null checks (#10219

### PR DESCRIPTION
**Pull request by t3chguy**
Requires https://github.com/matrix-org/matrix-js-sdk/pull/3175

---
This change is marked as an *internal change* (Task), so will not be included in the changelog.

### CODE REVIEW

File: src/components/views/auth/PasswordLogin.tsx [hunk 1]
```diff
@@ -116,25 +116,25 @@ export default class PasswordLogin extends React.PureComponent<IProps, IState> {
     };
 
     private onUsernameChanged = (ev: React.ChangeEvent<HTMLInputElement>): void => {
-        this.props.onUsernameChanged(ev.target.value);
+        this.props.onUsernameChanged!(ev.target.value);
     };
 
     private onUsernameBlur = (ev: React.FocusEvent<HTMLInputElement>): void => {
-        this.props.onUsernameBlur(ev.target.value);
+        this.props.onUsernameBlur!(ev.target.value);
```
Comment (weeman1337) [inline src/components/views/auth/PasswordLogin.tsx:123]: Should there not better be a `?` used here? Per props it is optional.

```suggestion
        this.props.onUsernameBlur?.(ev.target.value);
```

(applies to other pro usages as well)

File: src/components/views/elements/Validation.tsx [hunk 2]
```diff
@@ -111,13 +108,13 @@ export default function withValidation<T = void, D = void>({
                     continue;
                 }
 
-                if (rule.skip?.call(this, data, derivedData)) {
+                if (rule.skip?.call(this, data, derivedData!)) {
```
Comment (weeman1337) [inline src/components/views/elements/Validation.tsx:111]: Why is `derivedData` set here?
Maybe in non-obvious cases we should add comments?

File: src/components/views/rooms/RoomListHeader.tsx [hunk 3]
```diff
@@ -367,7 +373,7 @@ const RoomListHeader: React.FC<IProps> = ({ onVisibilityChange }) => {
 
     let title: string;
     if (activeSpace) {
-        title = spaceName;
+        title = spaceName!;
```
Comment (weeman1337) [inline src/components/views/rooms/RoomListHeader.tsx:376]: In cases like these should we just check the variable to be set?
Here for instance

```
if (activeSpace && spaceName) {
```

**Review by weeman1337 — APPROVED** [review]
:+1: some smaller notes

File: src/components/views/auth/PasswordLogin.tsx (hunk 1 above)
Comment (t3chguy) [reply src/components/views/auth/PasswordLogin.tsx:123]: `onUsernameBlur` is only ever used if `this.props.onUsernameBlur` is set, it'd be a lot cleaner as an inline anonymous function but we don't like those in class components, so the null-guard generated by TS is redundant

File: src/components/views/elements/Validation.tsx (hunk 2 above)
Comment (t3chguy) [reply src/components/views/elements/Validation.tsx:111]: The issue is actually higher up, `derivedData` will always be of the right type, its just that the type is sometimes `undefined` - I couldn't find an ideal way to express it in the generic. If `deriveData === undefined then D is undefined, else D is something else`

File: src/components/views/rooms/RoomListHeader.tsx (hunk 3 above)
Comment (t3chguy) [reply src/components/views/rooms/RoomListHeader.tsx:376]: Perhaps, this is a case where we know if activeSpace is set that we also set spaceName, TS doesn't really have a sane way of expressing these relationships

File: src/components/views/elements/Validation.tsx (hunk 2 above)
Comment (t3chguy) [reply src/components/views/elements/Validation.tsx:111]: Can do it with a cast on the set instead, much better

**Review by justjanne — APPROVED** [review]

**Pull request merged**